File: src/app/LibraryModel.cpp
#include "LibraryModel.h"

#include "core/SettingsStore.h"

#include <QDateTime>
#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QUrl>

LibraryModel::LibraryModel(SettingsStore *settings, QObject *parent)
    : QAbstractListModel(parent)
    , m_settings(settings)
{
    connect(settings, &SettingsStore::projectsDirChanged, this, &LibraryModel::refresh);
    refresh();
}

int LibraryModel::rowCount(const QModelIndex &parent) const
{
    return parent.isValid() ? 0 : int(m_projects.size());
}

QVariant LibraryModel::data(const QModelIndex &index, int role) const
{
    if (!index.isValid() || index.row() < 0 || index.row() >= m_projects.size())
        return {};

    const Project &p = m_projects.at(index.row());
    switch (role) {
    case NameRole:
        return p.folderName;
    case ProductRole:
        return p.productName;
    case HookRole:
        return p.hook;
    case CreatedRole:
        return p.createdAt;
    case DirRole:
        return p.dir;
    case FinalVideoRole:
        return p.finalVideo;
    case ThumbnailRole:
        return p.thumbnail.isEmpty() ? QString() : QUrl::fromLocalFile(p.thumbnail).toString();
    case SuccessRole:
        return p.success;
    case CostRole:
        return p.cost;
    case HasCostRole:
        return p.hasCost;
    default:
        return {};
    }
}

double LibraryModel::totalCost() const
{
    double total = 0.0;
    for (const Project &project : m_projects)
        total += project.cost;
    return total;
}

QHash<int, QByteArray> LibraryModel::roleNames() const
{
    return {
        {NameRole, "name"},         {ProductRole, "productName"},
        {HookRole, "hook"},         {CreatedRole, "createdAt"},
        {DirRole, "dir"},           {FinalVideoRole, "finalVideo"},
        {ThumbnailRole, "thumbnail"}, {SuccessRole, "success"},
        {CostRole, "cost"},         {HasCostRole, "hasCost"},
    };
}

void LibraryModel::refresh()
{
    beginResetModel();
    m_projects.clear();

    QDir root(m_settings->projectsDir());
    if (root.exists()) {
        const QFileInfoList folders =
            root.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot, QDir::Name | QDir::Reversed);

        for (const QFileInfo &folder : folders) {
            QFile manifestFile(QDir(folder.absoluteFilePath()).filePath(QStringLiteral("project.json")));
            if (!manifestFile.open(QIODevice::ReadOnly))
                continue;

            const QJsonObject manifest = QJsonDocument::fromJson(manifestFile.readAll()).object();

            Project project;
            project.folderName = folder.fileName();
            project.dir = folder.absoluteFilePath();
            project.productName = manifest.value(QStringLiteral("productName")).toString();
            project.hook = manifest.value(QStringLiteral("hook")).toString();
            project.success = manifest.value(QStringLiteral("success")).toBool();

            // Absent in projects made before costs were recorded: show nothing
            // rather than a confident "$0".
            if (manifest.contains(QStringLiteral("cost"))) {
                const QJsonObject cost = manifest.value(QStringLiteral("cost")).toObject();
                project.cost = cost.value(QStringLiteral("total")).toDouble();
                project.hasCost = !cost.value(QStringLiteral("lines")).toArray().isEmpty();
            }

            const QDateTime created =
                QDateTime::fromString(manifest.value(QStringLiteral("createdAt")).toString(),
                                      Qt::ISODate);
            project.createdAt = created.isValid()
                ? created.toString(QStringLiteral("d MMM yyyy, HH:mm"))
                : folder.birthTime().toString(QStringLiteral("d MMM yyyy, HH:mm"));

            const auto resolve = [&folder](const QString &relative) {
                return relative.isEmpty()
                    ? QString()
                    : QDir(folder.absoluteFilePath()).filePath(relative);
            };
            project.finalVideo = resolve(manifest.value(QStringLiteral("final")).toString());
            project.thumbnail = resolve(manifest.value(QStringLiteral("frame")).toString());

            if (!project.finalVideo.isEmpty() && !QFileInfo::exists(project.finalVideo))
                project.finalVideo.clear();
            if (!project.thumbnail.isEmpty() && !QFileInfo::exists(project.thumbnail))
                project.thumbnail.clear();

            m_projects.append(project);
        }
    }

    endResetModel();
    emit countChanged();
}

File: market_queen_flutter_app/lib/models/library_model.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:intl/intl.dart';
import 'package:path/path.dart' as p;

import '../core/settings_store.dart';

class LibraryProject {
  LibraryProject({
    required this.folderName,
    required this.productName,
    required this.hook,
    required this.createdAt,
    required this.dir,
    required this.finalVideo,
    required this.thumbnail,
    required this.success,
    required this.cost,
    required this.hasCost,
  });

  final String folderName;
  final String productName;
  final String hook;
  final String createdAt;
  final String dir;
  final String finalVideo;
  final String thumbnail;
  final bool success;
  final double cost;
  final bool hasCost;
}

/// Every run leaves a folder behind with a project.json manifest; the library is
/// just a listing of those folders, newest first.
class LibraryModel extends ChangeNotifier {
  LibraryModel(this._settings) {
    _settings.addListener(_onSettingsChanged);
    refresh();
  }

  final SettingsStore _settings;
  final List<LibraryProject> _projects = [];

  String _watchedDir = '';

  List<LibraryProject> get projects => List.unmodifiable(_projects);
  int get count => _projects.length;

  /// What every listed project cost, added up. Projects written before costs
  /// were recorded contribute nothing.
  double get totalCost =>
      _projects.fold(0.0, (total, project) => total + project.cost);

  @override
  void dispose() {
    _settings.removeListener(_onSettingsChanged);
    super.dispose();
  }

  void _onSettingsChanged() {
    if (_settings.projectsDir == _watchedDir) return;
    refresh();
  }

  void refresh() {
    _watchedDir = _settings.projectsDir;
    _projects.clear();

    final root = Directory(_watchedDir);
    if (root.existsSync()) {
      final folders = root
          .listSync()
          .whereType<Directory>()
          .toList()
        // Newest first: the folder names start with a timestamp.
        ..sort((a, b) => p.basename(b.path).compareTo(p.basename(a.path)));

      for (final folder in folders) {
        final project = _read(folder);
        if (project != null) _projects.add(project);
      }
    }

    notifyListeners();
  }

  LibraryProject? _read(Directory folder) {
    final manifestFile = File(p.join(folder.path, 'project.json'));
    if (!manifestFile.existsSync()) return null;

    Object? decoded;
    try {
      decoded = jsonDecode(manifestFile.readAsStringSync());
    } on Exception {
      return null;
    }
    if (decoded is! Map<String, dynamic>) return null;

    // Absent in projects made before costs were recorded: show nothing rather
    // than a confident "$0".
    var cost = 0.0;
    var hasCost = false;
    final costEntry = decoded['cost'];
    if (costEntry is Map) {
      cost = (costEntry['total'] as num?)?.toDouble() ?? 0.0;
      final lines = costEntry['lines'];
      hasCost = lines is List && lines.isNotEmpty;
    }

    final created = DateTime.tryParse('${decoded['createdAt'] ?? ''}') ??
        folder.statSync().changed;

    String resolve(Object? relative) {
      final value = '${relative ?? ''}';
      if (value.isEmpty) return '';
      final full = p.join(folder.path, value);
      return File(full).existsSync() ? full : '';
    }

    return LibraryProject(
      folderName: p.basename(folder.path),
      dir: folder.path,
      productName: '${decoded['productName'] ?? ''}',
      hook: '${decoded['hook'] ?? ''}',
      success: decoded['success'] == true,
      cost: cost,
      hasCost: hasCost,
      createdAt: DateFormat('d MMM yyyy, HH:mm').format(created),
      finalVideo: resolve(decoded['final']),
      thumbnail: resolve(decoded['frame']),
    );
  }
}
